staging/src/k8s.io/kube-openapi/pkg/handler/handler.go

```diff
@@ -0,0 +1,188 @@
+
+	specBytes []byte
+	specPb    []byte
+	specPbGz  []byte
+
+	specBytesETag string
+	specPbETag    string
+	specPbGzETag  string
+}
+
+func computeETag(data []byte) string {
+	return fmt.Sprintf("\"%X\"", sha512.Sum512(data))
+}
+
+// BuildAndRegisterOpenAPIService first build and spec and then registers a handler to provides access to it.
+// Use this method if your OpenAPI spec is static. If you want to update the spec, use BuildOpenAPISpec then RegisterOpenAPIService.
+func BuildAndRegisterOpenAPIService(servePath string, webServices []*restful.WebService, config *common.Config, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+	spec, err := builder.BuildOpenAPISpec(webServices, config)
+	if err != nil {
+		return nil, err
+	}
+	return RegisterOpenAPIService(spec, servePath, mux)
+}
+
+// RegisterOpenAPIService registers a handler to provides access to provided swagger spec.
+func RegisterOpenAPIService(openapiSpec *spec.Swagger, servePath string, mux *genericmux.PathRecorderMux) (*OpenAPIService, error) {
+	if !strings.HasSuffix(servePath, JsonExt) {
+		return nil, fmt.Errorf("Serving path must ends with \"%s\".", JsonExt)
+	}
+
+	servePathBase := strings.TrimSuffix(servePath, JsonExt)
+
+	o := OpenAPIService{}
+	if err := o.UpdateSpec(openapiSpec); err != nil {
+		return nil, err
+	}
+
+	mime.AddExtensionType(".json", MimeJson)
+	mime.AddExtensionType(".pb-v1", MimePb)
+	mime.AddExtensionType(".gz", MimePbGz)
+
+	type fileInfo struct {
+		ext            string
+		getDataAndETag func() ([]byte, string)
+	}
+
+	files := []fileInfo{
+		{".json", o.getSwaggerBytes},
+		{"-2.0.0.json", o.getSwaggerBytes},
+		{"-2.0.0.pb-v1", o.getSwaggerPbBytes},
+		{"-2.0.0.pb-v1.gz", o.getSwaggerPbGzBytes},
+	}
+
+	for _, file := range files {
+		path := servePathBase + file.ext
+		getDataAndETag := file.getDataAndETag
+		mux.Handle(path, gziphandler.GzipHandler(http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				if r.URL.Path != path {
+					w.WriteHeader(http.StatusNotFound)
+					w.Write([]byte("Path not found!"))
+					return
+				}
+				defer o.mu.Unlock()
+				o.mu.Lock()
+				data, etag := getDataAndETag()
+				w.Header().Set("Etag", etag)
+
+				// ServeContent will take care of caching using eTag.
+				http.ServeContent(w, r, path, o.lastModified, bytes.NewReader(data))
+			}),
+		))
+	}
+
+	return &o, nil
+}
+
+func (o *OpenAPIService) getSwaggerBytes() ([]byte, string) {
+	return o.specBytes, o.specBytesETag
+}
+
+func (o *OpenAPIService) getSwaggerPbBytes() ([]byte, string) {
+	return o.specPb, o.specPbETag
+}
+
+func (o *OpenAPIService) getSwaggerPbGzBytes() ([]byte, string) {
+	return o.specPbGz, o.specPbGzETag
+}
+
+func (o *OpenAPIService) UpdateSpec(openapiSpec *spec.Swagger) (err error) {
+	defer o.mu.Unlock()
+	o.mu.Lock()
+
+	o.orgSpec = openapiSpec
+	o.specBytes, err = json.MarshalIndent(openapiSpec, " ", " ")
+	if err != nil {
+		return err
+	}
+	o.specPb, err = toProtoBinary(o.specBytes)
```

Comment:
all these lines will render the object invalid because later accesses to the nil members will possibly panic. Would be better to make the whole modification atomic.